primitives/phragmen/src/lib.rs

```diff
@@ -340,38 +395,88 @@ pub fn elect<AccountId, Balance, FS, C>(
-	elected_stashes: &Vec<AccountId>,
-	assignments: &Vec<(AccountId, Vec<PhragmenAssignment<AccountId>>)>,
-	stake_of: FS,
-) -> SupportMap<AccountId> where
+/// Build the support map from the given phragmen result. It maps a flat structure like
+///
+/// ```nocompile
+/// assignments: vec![
+/// 	voter1, vec![(candidate1, w11), (candidate2, w12)],
+/// 	voter2, vec![(candidate1, w21), (candidate2, w22)]
+/// ]
+/// ```
+///
+/// into a mapping of candidates and their respective support:
+///
+/// ```nocompile
+///  SupportMap {
+/// 	candidate1: Support {
+/// 		own:0,
+/// 		total: w11 + w21,
+/// 		others: vec![(candidate1, w11), (candidate2, w21)]
+///		},
+/// 	candidate2: Support {
+/// 		own:0,
+/// 		total: w12 + w22,
+/// 		others: vec![(candidate1, w12), (candidate2, w22)]
+///		},
+/// }
+/// ```
+/// The second returned flag indicates the number of edges who corresponded to an actual winner from
+/// the given winner set. A value in this place larger than 0 indicates a potentially faulty
+/// assignment.
+///
+/// `O(E)` where `E` is the total number of edges.
+pub fn build_support_map<Balance, AccountId>(
+	winners: &Vec<AccountId>,
+	assignments: &Vec<StakedAssignment<AccountId>>,
+) -> (SupportMap<AccountId>, u32) where
 	AccountId: Default + Ord + Member,
 	Balance: Default + Copy + SimpleArithmetic,
-	C: Convert<Balance, u64> + Convert<u128, Balance>,
-	for<'r> FS: Fn(&'r AccountId) -> Balance,
 {
-	let to_votes = |b: Balance| <C as Convert<Balance, u64>>::convert(b) as ExtendedBalance;
+	let mut errors = 0;
 	// Initialize the support of each candidate.
 	let mut supports = <SupportMap<AccountId>>::new();
-	elected_stashes
+	winners
 		.iter()
 		.for_each(|e| { supports.insert(e.clone(), Default::default()); });
 
 	// build support struct.
-	for (n, assignment) in assignments.iter() {
-		for (c, per_thing) in assignment.iter() {
-			let nominator_stake = to_votes(stake_of(n));
-			// AUDIT: it is crucially important for the `Mul` implementation of all
-			// per-things to be sound.
-			let other_stake = *per_thing * nominator_stake;
+	for StakedAssignment { who, distribution } in assignments.iter() {
+		for (c, weight_extended) in distribution.iter() {
 			if let Some(support) = supports.get_mut(c) {
-				support.voters.push((n.clone(), other_stake));
-				support.total = support.total.saturating_add(other_stake);
+				support.total = support.total.saturating_add(*weight_extended);
+				support.voters.push((who.clone(), *weight_extended));
+			} else {
+				errors = errors.saturating_add(1);
 			}
 		}
 	}
-	supports
+	(supports, errors)
+}
+
+/// Evaluate a phragmen result, given the support map. The returned tuple contains:
+///
+/// - Minimum support. This value must be **maximized**.
+/// - Sum of all supports. This value must be **maximized**.
+/// - Sum of all supports squared. This value must be **minimized**.
+///
+/// `O(E)` where `E` is the total number of edges.
+pub fn evaluate_support<AccountId>(
```

Comment:
I think they each convey a different objective. The third one, namely, is the _least_ important one since it is afaik representing the degree of variance, which want to be minimised. This is, for example, a different score parameter than the second one which requires the total stake to be maximised.

Reply:
Oh yes it does not have the same meaning. I wonder if there could be a more appropriate evalution criteria.
My way of doing thing here would be to use a few 100% ratio, as unreliable as possible validators to minimize the chance the first value applies to my more relevant validator, then it will be the third condition that would really matter in the solution leading to major validator aligning to the low variance. 
Hum I would need to give it more thought, but generally I like the third condition better as it takes in account the full set.
Maybe first thing make sense because we did use phragmen algorithm, but in this case we evaluate a solution, I doubt it is good.

Reply:
Please forget about latest message, I did not convince myself the third one does not imply the first one, I will need to find a case where minimum degree of variance is possible with a smallest support non maximized: this seems rather counter intuitive (smallest support is bigger contributor in lowering variance by being maximized so intuitively I would think third condition implies first one).

Reply:
Resolving, since I got my counter example, still it seems that either comparing variance only or successive minimal value (replace variance by sort) would be more consistent.
With current criterium maybe calculating variance in a second loop (only if the minimums are not equals) is better.

Reply:
I'm also sure @alfonso in web3 will be happy to further discuss this. You seem to have good critics about this.